### PR TITLE
Proposed improvements to the early callback API.

### PR DESCRIPTION
**Pull request by davidben**
This proposes a couple improvements to `SSL_CTX_set_early_cb`:

* Now that early data exists, the name "early callback" is a little ambiguous. I propose "ClientHello callback".
* The early callback has a tri-state return with magic numbers. Define some constants instead.

Thoughts? (The API has not been released yet, so I am assuming it is not yet set in stone.)

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
The travis failure is relevant. `make doc-nits` is complaining.

File: test/sslapitest.c [hunk 1]
```diff
@@ -422,26 +422,27 @@ static int full_early_callback(SSL *s, int *al, void *arg)
...
-        return 0;
-    if (!SSL_early_get1_extensions_present(s, &exts, &len))
-        return 0;
+        return SSL_CLIENT_HELLO_ERROR;
+    if (!SSL_client_hello_get1_extensions_present(s, &exts, &len))
+        return SSL_CLIENT_HELLO_ERROR;
     if (len != OSSL_NELEM(expected_extensions) ||
         memcmp(exts, expected_extensions, len * sizeof(*exts)) != 0) {
-        printf("Early callback expected ClientHello extensions mismatch\n");
+        printf("ClientHello callback expected ClientHello extensions mismatch\n");
```
Comment (kaduk) [inline test/sslapitest.c]: This line now comes across as quite banal, but as it's mostly just a unique identifier for where in the source the test is failing, it doesn't really matter.

File: doc/man3/SSL_CTX_set_client_hello_cb.pod [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+=head1 NAME
+
+SSL_CTX_set_client_hello_cb, SSL_client_hello_cb_fn, SSL_client_hello_isv2, SSL_client_hello_get0_legacy_version, SSL_client_hello_get0_random, SSL_client_hello_get0_session_id, SSL_client_hello_get0_ciphers, SSL_client_hello_get0_compression_methods, SSL_client_hello_get1_extensions_present, SSL_client_hello_get0_ext - callback functions for early server-side ClientHello processing
+
+=head1 SYNOPSIS
+
+ #define SSL_CLIENT_HELLO_SUCCESS 1
+ #define SSL_CLIENT_HELLO_ERROR   0
+ #define SSL_CLIENT_HELLO_RETRY   (-1)
```
Comment (kaduk) [inline doc/man3/SSL_CTX_set_client_hello_cb.pod]: These three need to go in the NAME section as well, per find-doc-nits.

File: test/sslapitest.c (hunk 1 above)
Comment (davidben) [reply test/sslapitest.c]: Switched it to just "expected extensions mismatch". I thought I did that already, but maybe that was somewhere else and I missed that one.

File: doc/man3/SSL_CTX_set_client_hello_cb.pod (hunk 2 above)
Comment (davidben) [reply doc/man3/SSL_CTX_set_client_hello_cb.pod]: It then complains that they're not public, which seems to be about the .num files. Other constants like `SSL_READ_EARLY_DATA_ERROR` are not listed there or at the top of .pod files, so it looks like it'll be more consistent to remove them from this section? I've done that and doc-nits is happy now.

Comment (kaduk) [reply doc/man3/SSL_CTX_set_client_hello_cb.pod]: util/private.num can contain entries for defines such as this that are public^Wdocumentable, if you can get your head around the naming of it.

Comment (richsalz) [reply doc/man3/SSL_CTX_set_client_hello_cb.pod]: Yeah, it's a bad name.  My fault.  But more importantly, our style is not to list specific constant values in the documentation, but instead to refer to just the names in the prose.

**Comment by richsalz** [comment]
reconfirm.

**Comment by kaduk** [comment]
I'll squash+merge

**Comment by kaduk** [comment]
merged; closing

**Pull request closed without merging**